paddle/fluid/framework/details/multi_devices_graph_builder.cc

```diff
@@ -396,30 +396,24 @@ void MultiDevSSAGraphBuilder::InsertDataBalanceOp(
   }
 }
 
-bool MultiDevSSAGraphBuilder::IsParameterGradientOnce(
-    const std::string &og,
-    std::unordered_set<std::string> *og_has_been_broadcast) const {
-  bool is_pg_once =
-      grad_names_.count(og) != 0 && og_has_been_broadcast->count(og) == 0;
-  if (is_pg_once) {
-    // Insert NCCL AllReduce Op
-    og_has_been_broadcast->insert(og);
-  }
-  return is_pg_once;
-}
-
 int MultiDevSSAGraphBuilder::GetOpDeviceID(const OpDesc &op) const {
   if (strategy_.reduce_ != BuildStrategy::ReduceStrategy::kReduce) {
     return -1;
   }
 
-  for (auto &varname : op.InputArgumentNames()) {
-    int dev_id = GetVarDeviceID(varname);
-    if (dev_id != -1) {
-      return dev_id;
-    }
+  int op_role = boost::get<int>(
+      op.GetAttr(framework::OpProtoAndCheckerMaker::OpRoleAttrName()));
+  if (op_role != static_cast<int>(framework::OpRole::kOptimize)) {
+    return -1;
   }
-  return -1;
+  auto &params = boost::get<std::vector<std::string>>(
+      op.GetAttr(framework::OpProtoAndCheckerMaker::OpRoleVarAttrName()));
+
+  PADDLE_ENFORCE_EQ(params.size(), 1U);
+  std::string grad = framework::GradVarName(params[0]);
+  int dev_id = GetVarDeviceID(grad);
```

Comment:
<img width="1280" alt="6ccfe2016af2a8ae4ba6cacc6" src="https://user-images.githubusercontent.com/30176695/42618903-ee68abae-85e8-11e8-849a-9f538f7a6f4b.png">

It is not appropriate, because in this case, both `adam`'s and `scale`'s role are `kOptimize`.